src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -413,13 +450,57 @@ public SolutionState RemoveProjects(ArrayBuilder<ProjectId> projectIds)
         foreach (var projectId in projectIds)
             newDependencyGraph = newDependencyGraph.WithProjectRemoved(projectId);
 
+        var languageCountDeltas = new TemporaryArray<(string language, int count)>();
+        foreach (var projectId in projectIds)
+        {
+            AddLanguageCount(ref languageCountDeltas, _projectIdToProjectStateMap[projectId].Language, amount: -1);
+        }
+
         return this.Branch(
             solutionAttributes: newSolutionAttributes,
             projectIds: newProjectIds,
             idToProjectStateMap: newStateMap,
+            projectCountByLanguage: AddLanguageCounts(ProjectCountByLanguage, languageCountDeltas),
             dependencyGraph: newDependencyGraph);
     }
 
+    private static void AddLanguageCount(ref TemporaryArray<(string language, int count)> languageCounts, string language, int amount)
+    {
+        var index = languageCounts.IndexOf(static (c, language) => c.language == language, language);
+        if (index < 0)
+        {
+            languageCounts.Add((language, amount));
+        }
+        else
+        {
+            languageCounts[index] = (language, languageCounts[index].count + amount);
+        }
+    }
+
+    private static ImmutableDictionary<string, int> AddLanguageCounts(ImmutableDictionary<string, int> projectCountByLanguage, in TemporaryArray<(string language, int count)> languageCountDeltas)
+    {
+        foreach (var (language, delta) in languageCountDeltas)
+        {
+            if (!projectCountByLanguage.TryGetValue(language, out var currentCount))
+            {
+                currentCount = 0;
+            }
+
+            var newCount = currentCount + delta;
+            if (newCount > 0)
+            {
+                projectCountByLanguage = projectCountByLanguage.SetItem(language, newCount);
+            }
+            else
+            {
+                Debug.Assert(newCount == 0);
```

Comment:
consider making a contract call.

Reply:
i now definitely want this to be a contract call.  going below 0 needs to be a hard break that we catch early :)